### PR TITLE
Hypothesis testing framework integrations for Polars 

### PR DESCRIPTION
**Pull request by alexander-beedie**
### What is Hypothesis?

TLDR: a powerful testing framework based on an idea popularised by the Haskell library [Quickcheck](https://hackage.haskell.org/package/QuickCheck) (back in the day!)

> [Hypothesis](https://hypothesis.works/) is a library for creating unit tests which are simpler to write and more powerful when run, finding edge cases in your code you wouldn’t have thought to look for. It is stable, powerful and easy to add to any existing test suite.

### Polars-specific integrations

I took a look at their [pandas strategy units](https://hypothesis.readthedocs.io/en/latest/numpy.html#pandas) and decided to write our own in a similar vein, covering DataFrame, LazyFrame, and Series, but with a richer set of options. This PR adds support for all three, along with test coverage for the new units (which double as a small working example of some of the features), additions to the testing docs, and a good set of informative / comprehensive docstrings.

### Show me the money, Jerry!

_**Obvious question**: "Is this yet another one of those clever-sounding things that is a pain to use?"_
_**Surprising answer**: "No." :)_

Now the strategy units exist, using them is easy; for example, here's a simple invocation of the DataFrames strategy:

```python
import polars as pl
from hypothesis import given
from polars.testing import dataframes

@given(df=dataframes())
def test_repr(df: pl.DataFrame) -> None:
    assert isinstance(repr(df), str)
    # print( df )
```

_**Likely response**: "I am incredibly underwhelmed - that just looks like a fixture" :p_

Which is true, but... uncomment the `print( df )` and you'll start to get a sense of what's going on, because you'll discover that the test actually executes across about 100 DataFrames (this is configurable) with wildly varying schemas, different sizes, and it actively probes at typical edge-case values (integer boundaries, weird strings, weird dates, etc), as well as scenarios like _"what if there is no data?"_, _"what if there is data, but it's all NULL?"_, _"what if there are no columns?"_, and so forth. 

And you get all that from applying a decorator.

_And_ this test found that some assumptions I had about datetime ranges were wrong, as it actually triggered a PanicException, so I have currently put a floor on date time-generation of 1970-01-01 while I work out whether it's a real bug or not! (So even that trivial test _may_ have uncovered something).

Instrumenting just **two** small tests this way has already found one "maybe" bug (above), and two _actual_ edge-case bugs (#3828, now fixed, and another that is fixed with this PR). The second bug: `assert_series_equal` actually fails if two identical _but empty_ Series are passed to it  :)

### Show me some more

I also instrumented the `null_count` test; it immediately failed on the empty frame scenario (no rows/cols) so I added an explicit check for that to the test (as it's expected, but wasn't covered). I then added a trivial assert/comparison to validate a simple python-based count against the internal implementation - and this is the code that actually found the `ZeroDivisionError` on `time(0,0,0)` values.
 
```python
@given(df=dataframes())
def test_null_count(df: pl.DataFrame) -> None:
    null_count, ncols = df.null_count(), len(df.columns)
    if ncols == 0:
        assert null_count.shape == (0, 0)
    else:
        assert null_count.shape == (1, ncols)
        for idx, count in enumerate(null_count.rows()[0]):
            assert count == sum(v is None for v in df.select_at_idx(idx).to_list())
```
How about LazyFrame and Series? 

```python
# generate/test LazyFrames
@given(ldf=dataframes(lazy=True))
def test_stuff(ldf: pl.LazyFrame):
    ...
```
```python
# generate/test Series
@given(s=series())
def test_stuff(s: pl.Series):
    ...
```
And what else can it do? If we start turning the dials up, with a few more options and use of the `column` and `column` helpers then we get things like this that offer more detailed control, but remain easy to understand and write (this is actually one of the new unit tests _for_ these new testing strategies):
```python
@given(
    lf=dataframes(
        # generate lazyframes with at least one row
        lazy=True,
        min_size=1,
        # test mix & match of bulk-assigned cols with custom cols
        cols=columns(["a", "b"], dtype=pl.UInt8, unique=True),
        include_cols=[
            column("c", dtype=pl.Boolean),
            column("d", strategy=sampled_from(["x", "y", "z"])),
        ],
    )
)
def test_strategy_frame_columns(lf: pl.LazyFrame) -> None:
    assert lf.schema == {"a": pl.UInt8, "b": pl.UInt8, "c": pl.Boolean, "d": pl.Utf8}
    assert lf.columns == ["a", "b", "c", "d"]
    df = lf.collect()

    # confirm uint cols bounds
    uint8_max = (2**8) - 1
    assert df["a"].min() >= 0
    assert df["b"].min() >= 0
    assert df["a"].max() <= uint8_max
    assert df["b"].max() <= uint8_max

    # confirm uint cols uniqueness
    assert df["a"].is_unique().all()
    assert df["a"].is_unique().all()

    # boolean col
    assert all(isinstance(v, bool) for v in df["c"].to_list())

    # string col, entries selected from custom values
    xyz = {"x", "y", "z"}
    assert all(v in xyz for v in df["d"].to_list())
```
Support is there for per-column custom data strategies, native hypothesis strategies can be freely mixed-in (eg: `sampled_from`, above), size constraints, type constraints/exclusions, etc.

### The end...

Plenty more in the inline docs, as well as some additional examples/tests, with more to come - hope it helps :)

### CODE REVIEW

**Comment by alexander-beedie** [comment]
**Misc**

Unrelated to the new test code, I also updated the `__hash__` of the temporal DataTypes to include the time unit. This does _not_ change their equality-checking behaviour, but _does_ now mean that they can all independently exist as entries in a set or as keys in a dict.

**Comment by alexander-beedie** [comment]
Huh. The same test went and found another error - this time with `date`, and seemingly only on Windows. I don't have access to a Windows machine at home, so I can't run the repro:

> pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: PyErr { type: <class 'OSError'>, value: OSError(22, 'Invalid argument'), traceback: Some(<traceback object at 0x0000021A02711600>) }
> [336](https://github.com/pola-rs/polars/runs/7099470684?check_suite_focus=true#step:6:337)
> C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\polars\internals\series.py:1993: PanicException
> [337](https://github.com/pola-rs/polars/runs/7099470684?check_suite_focus=true#step:6:338)
> ---------------------------- Captured stderr call -----------------------------
> [338](https://github.com/pola-rs/polars/runs/7099470684?check_suite_focus=true#step:6:339)
> thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: PyErr { type: <class 'OSError'>, value: OSError(22, 'Invalid argument'), traceback: Some(<traceback object at 0x0000021A02711600>) }', src\conversion.rs:516:74
> [339](https://github.com/pola-rs/polars/runs/7099470684?check_suite_focus=true#step:6:340)
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> [340](https://github.com/pola-rs/polars/runs/7099470684?check_suite_focus=true#step:6:341)
> --------------------------------- Hypothesis ----------------------------------
> [341](https://github.com/pola-rs/polars/runs/7099470684?check_suite_focus=true#step:6:342)
> You can add `@seed(218914631208369795481727013752704885687)` to this test or run pytest with `--hypothesis-seed=218914631208369795481727013752704885687` to reproduce this failure.

I've temporarily comment-out the test so that the PR can succeed and get merged  ;)

File: py-polars/polars/datatypes.py [hunk 1]
```diff
@@ -407,19 +425,18 @@ def __hash__(self) -> int:
...
-        (Duration, "us"): pa.duration("us"),
-        (Duration, "ns"): pa.duration("ns"),
+        Duration("ms"): pa.duration("ms"),
+        Duration("us"): pa.duration("us"),
+        Duration("ns"): pa.duration("ns"),
         Time: pa.time64("us"),
-        (Time, "ms"): pa.time32("ms"),
-        (Time, "us"): pa.time64("us"),
-        (Time, "ns"): pa.time64("ns"),
+        # Time("ms"): pa.time32("ms"),
```
Comment (ritchie46) [inline py-polars/polars/datatypes.py:437]: Was this intended?

**Comment by ritchie46** [comment]
Thanks a lot @alexander-beedie. I still have to let it all sink in a bit. But I guess that will come when working with it. I believe that this can catch more edge cases and that's definitely a huge plus as there are a lot.

File: py-polars/polars/datatypes.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/datatypes.py:437]: Yes - we don't actually support timeunits for the `Time` type; I must have added them before in anticipation of such a development, but as `Time(tu)` can't be declared, it can't exist in the lookup dict. If support gets added int he future, can reinstate.
